src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,523 @@
+
+/*
+ * Copyright (C) cuiweixie
+ */
+
+#ifndef NGX_LUA_NO_FFI_API
+
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_util.h"
+#include "ngx_http_lua_semaphore.h"
+#include "ngx_http_lua_contentby.h"
+
+
+ngx_int_t ngx_http_lua_semaphore_init_mm(ngx_http_lua_semaphore_mm_t *mm);
+static ngx_http_lua_semaphore_t *ngx_http_lua_alloc_semaphore(void);
+void ngx_http_lua_cleanup_semaphore_mm(void *data);
+static void ngx_http_lua_free_semaphore(ngx_http_lua_semaphore_t *sem);
+static ngx_int_t ngx_http_lua_semaphore_resume(ngx_http_request_t *r);
+int ngx_http_lua_ffi_semaphore_new(ngx_http_lua_semaphore_t **psem,
+    int n, char **errstr);
+int ngx_http_lua_ffi_semaphore_post(ngx_http_lua_semaphore_t *sem,
+    int n, char **errstr);
+int ngx_http_lua_ffi_semaphore_wait(ngx_http_request_t *r,
+    ngx_http_lua_semaphore_t *sem, int wait_ms, u_char *errstr, size_t *errlen);
+static void ngx_http_lua_semaphore_cleanup(void *data);
+static void ngx_http_lua_semaphore_handler(ngx_event_t *ev);
+static void ngx_http_lua_semaphore_timeout_handler(ngx_event_t *ev);
+void ngx_http_lua_ffi_semaphore_gc(ngx_http_lua_semaphore_t *sem);
+
+
+enum {
+    SEMAPHORE_WAIT_SUCC = 0,
+    SEMAPHORE_WAIT_TIMEOUT = 1
+};
+
+
+static ngx_http_lua_semaphore_t *
+ngx_http_lua_alloc_semaphore(void)
+{
+    ngx_http_lua_semaphore_t          *sem;
+    ngx_http_lua_semaphore_t          *iter;
+    ngx_http_lua_main_conf_t          *lmcf;
+    ngx_queue_t                       *q;
+    ngx_uint_t                         i, n;
+    ngx_http_lua_semaphore_mm_block_t *block;
+    ngx_http_lua_semaphore_mm_t       *mm;
+
+    if (!ngx_cycle || !ngx_cycle->conf_ctx) {
+        dd("in ngx_http_lua_alloc_semaphore"
+           "ngx_cycle or ngx_cycle->conf_ctx is null");
+
+        return NULL;
+    }
+
+    lmcf = ngx_http_cycle_get_module_main_conf(ngx_cycle,
+                                               ngx_http_lua_module);
+
+    if (!lmcf) {
```

Comment:
Style: we usually test against `NULL` for pointers, as in

```C
if (lmcf == NULL) {
    ...
}
```

But I wonder what contexts you're protecting against here.